language/move-vm/runtime/src/interpreter.rs

```diff
@@ -1000,21 +1025,44 @@ impl Frame {
                             interpreter.operand_stack.last_n(field_count as usize)?,
                         )?;
                         let args = interpreter.operand_stack.popn(field_count)?;
-                        interpreter
-                            .operand_stack
-                            .push(Value::struct_(Struct::pack(args)))?;
+                        let value = Value::struct_(Struct::pack_with_tag(
+                            args,
+                            resolver
+                                .instantiate_generic_type(*si_idx, self.ty_args())?
+                                .get_hash(),
+                        ));
+
+                        interpreter.operand_stack.push(value)?;
                     }
-                    Bytecode::Unpack(_sd_idx) => {
+                    Bytecode::Unpack(sd_idx) => {
                         let struct_ = interpreter.operand_stack.pop_as::<Struct>()?;
+                        let ty = resolver.get_struct_type(*sd_idx);
+                        if let Some(ty1) = struct_.tag() {
+                            if ty1 != ty.get_hash() {
+                                return Err(PartialVMError::new(
+                                    StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR,
+                                )
+                                .with_message(format!("unexpected type mismatch")));
```

Comment:
I think more detail in these messages (e.g., name of the instruction in question, the two types) would be helpful.

Reply:
Done!